### PR TITLE
Refactored inner hits parsing and intoduced InnerHitBuilder

### PR DESCRIPTION
**Pull request by martijnvg**
Both top level and inline inner hits are now covered by InnerHitBuilder.
Although there are differences between top level and inline inner hits,
they now make use of the same builder logic.

The parsing of top level inner hits slightly changed to be more readable.
Before the nested path or parent/child type had to be specified as encapsuting
json object, now these settings are simple fields. Before this was required
to allow streaming parsing of inner hits without missing contextual information.

Once some issues are fixed with inline inner hits (around multi level hierachy of inner hits),
top level inner hits will be deprecated and removed in the next major version.

Also I think some `SearchParseElement` implementations can be removed, but I haven't done that yet.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/query/support/InnerHitsBuilder.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+    public final static InnerHitsBuilder PROTO = new InnerHitsBuilder(Collections.emptyMap());
+
+    private final Map<String, InnerHitBuilder> innerHitsBuilders;
+
+    public InnerHitsBuilder() {
+        this.innerHitsBuilders = new HashMap<>();
+    }
+
+    public InnerHitsBuilder(Map<String, InnerHitBuilder> innerHitsBuilders) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitsBuilder.java]: Maybe this can be private.

File: core/src/main/java/org/elasticsearch/index/query/support/InnerHitsBuilder.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        Map<String, InnerHitBuilder> innerHitBuilders = new HashMap<>();
+        for (Token token = parser.nextToken(); token != Token.END_OBJECT; token = parser.nextToken()) {
+            switch (token) {
+                case START_OBJECT:
+                    String name = parser.currentName();
+                    InnerHitBuilder innerHitBuilder = InnerHitBuilder.fromXContent(parser, context);
+                    innerHitBuilder.setName(name);
+                    innerHitBuilders.put(name, innerHitBuilder);
+                    break;
+            }
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitsBuilder.java]: Can we throw ParseExceptions on unexpected tokens here, e.g. when we encounter arrays or find anything that is not a START_OBJECT after a field name?

File: core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java [hunk 3]
```diff
@@ -0,0 +1,603 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(name, nestedPath, parentChildType, from, size, explain, version, trackScores, fieldNames,
+                fieldDataFields, scriptFields, fetchSourceContext, sorts, highlightBuilder, query, innerHitsBuilder);
+    }
+
+    @Override
+    public String toString() {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java]: If instead of implementing ToXContent we extend ToXContentToBytes, we get this from there. Don't know if this would work here.

File: core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java [hunk 4]
```diff
@@ -1239,7 +1233,7 @@ public SearchSourceBuilder readFrom(StreamInput in) throws IOException {
             builder.queryBuilder = in.readQuery();
         }
         if (in.readBoolean()) {
-            int size = in.readVInt();
+            size = in.readVInt();
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java]: This (and the following occurences) looks like its overwriting the global size field of the SearchSourceBuilder now. That's not the builder that is returned, so it might not matter but I find it very confusing. So I'd suggest also renaming the local variable to avoid shadowing the global name.

File: core/src/test/java/org/elasticsearch/index/query/HasChildQueryBuilderTests.java [hunk 5]
```diff
@@ -118,11 +116,11 @@ public IndexFieldDataService fieldData() {
     protected HasChildQueryBuilder doCreateTestQueryBuilder() {
         int min = randomIntBetween(0, Integer.MAX_VALUE / 2);
         int max = randomIntBetween(min, Integer.MAX_VALUE);
-        InnerHitsBuilder.InnerHit innerHit = new InnerHitsBuilder.InnerHit().setSize(100).addSort(STRING_FIELD_NAME_2, SortOrder.ASC);
         return new HasChildQueryBuilder(CHILD_TYPE,
                 RandomQueryBuilder.createQuery(random()), max, min,
                 RandomPicks.randomFrom(random(), ScoreMode.values()),
-                randomBoolean()  ? null : new QueryInnerHits("inner_hits_name", innerHit));
+                randomBoolean()  ? null : new InnerHitBuilder().setName("inner_hits_name").setSize(100)
```
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/index/query/HasChildQueryBuilderTests.java]: Can we randomize one or two properties of the InnerHitBuilder, e.g. name or size?

File: core/src/test/java/org/elasticsearch/index/query/HasParentQueryBuilderTests.java [hunk 6]
```diff
@@ -104,10 +104,10 @@ public IndexFieldDataService fieldData() {
      */
     @Override
     protected HasParentQueryBuilder doCreateTestQueryBuilder() {
-        InnerHitsBuilder.InnerHit innerHit = new InnerHitsBuilder.InnerHit().setSize(100).addSort(STRING_FIELD_NAME_2, SortOrder.ASC);
         return new HasParentQueryBuilder(PARENT_TYPE,
                 RandomQueryBuilder.createQuery(random()),randomBoolean(),
-                randomBoolean() ? null : new QueryInnerHits("inner_hits_name", innerHit));
+                randomBoolean() ? null : new InnerHitBuilder().setName("inner_hits_name").setSize(100)
```
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/index/query/HasParentQueryBuilderTests.java]: Same here, a little bit of randomness would be nice.

File: core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java [hunk 7]
```diff
@@ -0,0 +1,603 @@
...
+        } else {
+            throw new IllegalStateException("Neither a nested or parent/child inner hit");
+        }
+        setupInnerHitsContext(context, innerHitsContext);
+        return innerHitsContext;
+    }
+
+    public InnerHitsContext.BaseInnerHits buildTopLevel(SearchContext parentSearchContext, QueryShardContext context,
+                                                        InnerHitsContext innerHitsContext) throws IOException {
+        if (nestedPath != null) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java]: This looks like InnerHitBuilder can have either `nestedPath` or `parentChildType` set but never both. Could those two settings then be mandatory constructor args, and we'd have two constructors, one for each flavour?

Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java]: +1

File: core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java (hunk 3 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java]: Ok, I thought that `ToXContentToBytes` was only used for things that we didn't have a builder for. Makes sense then to extend `ToXContentToBytes`.

File: core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java [hunk 8]
```diff
@@ -0,0 +1,603 @@
...
+        this.query = query;
+        return this;
+    }
+
+    public InnerHitBuilder setInnerHitsBuilder(InnerHitsBuilder innerHitsBuilder) {
+        this.innerHitsBuilder = innerHitsBuilder;
+        return this;
+    }
+
+    public InnerHitsContext.BaseInnerHits buildInline(SearchContext parentSearchContext, QueryShardContext context) throws IOException {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java]: I was wondering what the difference between `buildInline` and `buildTolLevel` was, maybe this could be best explained in some javadoc. Other than that the two methods look quiet similar, not sure if the differences could be factored out, depends on whether this increases complexity a lot.

Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java]: The main differences are that top level inner hits can have its own query and allows for nesting. (so in case of a parent, child and grand_child scenario these can be nested inside a top level inner hits definition)

The reason I have made two different methods is that I didn't want inline inner hits logic to be complex (esp. the recursion that it has) while that isn't needed. Also top inner hits will be deprecated in 5.0 and removed later, which means this entire `buildTopLevel(...)` can simply be dropped. 

I'll add jdocs that explain this.

File: core/src/test/java/org/elasticsearch/index/query/support/InnerHitBuilderTests.java [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+                break;
+            case 3:
+                copy.setVersion(!copy.isVersion());
+                break;
+            case 4:
+                copy.setTrackScores(!copy.isTrackScores());
+                break;
+            case 5:
+                copy.setName("CHANGED");
+                break;
```
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/index/query/support/InnerHitBuilderTests.java]: Can this be expanded to a few more (if not all) potential properties of InnerHitBuilder? Maybe the more complex random generation of the parts can be factored out from randomInnerHits(). Theres also a helper in ESTestCase#randomValueOtherThan() that might help not accidentally picking the same value that was used in the original InnerHitBuilder.

File: core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java (hunk 3 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java]: I didn't know about that restriction, I only saw that we extend it in AbstractQueryBuilder, AggregatorBuilder etc..., other than that no strong opinion.

File: core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java (hunk 4 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java]: This also seems to let `org.elasticsearch.messy.tests.SearchStatsTests.testSimpleStats` fail consistently for me when running locally (`Dtests.seed=A8EFE9FB672752E6`)

**Comment by cbuescher** [comment]
@martijnvg this looks great, even though I didn't compare it completely to the way it was before the refactoring, I have the feeling this change simplifies a lot. Left a few minor comments, only thing that was causing some trouble with the tests on my side was the `size` shadowing issue in `SearchSourceBuilder#readFrom()`, but will run the whole tests again now with a local fix. Other than that, maybe it would be good to also have a basic test for `InnerHitsBuilder`, although that one doesn't have much logic of its own.
Its also a quiet a big PR, maybe somebody else involved in the search refactoring should take a second look after this round. I'll ask around.

File: core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java (hunk 4 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java]: ouch that is messy, I'll use different variable names.

Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java]: I'll revert these lines.

File: core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java (hunk 7 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java]: I tried to make this change, but it isn't really possible to have two constructors having both a single string parameter. Also because the object parser is used to parse the xcontent into a builder, no constructor arguments can be passed. The object parser can only work with setters.

Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java]: I like to keep the object parser here, because that is much easier to read than the regular xcontent parsing (with loops and such).

**Comment by martijnvg** [comment]
@cbuescher Thanks for looking at this. I updated the PR based on your comment and fixed the `size` shadowing issue.

**Comment by dimfeld** [comment]
With top-level `inner_hits` going away, is there a recommended way to replace the functionality? My usual case here involves using a `has_child` in the query to look for parents with a particular value in a child, and then a top-level `inner_hits` to return the top few child objects of the same type, ranked by some value using a `function_score`. 

I'm thinking that something like this will work; is there a better way?
```
{  query:  {  bool: {
   must: // main query with "real" child filter and whatever else goes here
   should: [
      { has_child: { type: childType, inner_hits: {} }
    ]
} } } 
```

**Comment by martijnvg** [comment]
@dimfeld Yes, that should work too. Also I'm thinking of also allowing the `query` option inside inline inner hits (which would then overwrite the actual inner query), so that you can do something like this:

```
{  query:  {  bool: {
   must: [
      { has_child: { type: childType, query: {real query}, inner_hits: {query: {function_score query}} } 
    ]
} } } 
```

This would be better, because it would be more efficient as only one `has_child` query would need to be specified. 

The goal with replacing inner hits is that same things can be done with inline inner hits as top level inner hits.

**Comment by dimfeld** [comment]
Sounds good Thanks @martijnvg!

**Comment by cbuescher** [comment]
@martijnvg thanks, changes look great and all works for me now, but maybe @colings86, @javanna or @s1monw want to take a second look at this before it gets in.

File: core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java [hunk 10]
```diff
@@ -0,0 +1,610 @@
...
+    public InnerHitBuilder setName(String name) {
+        this.name = name;
+        return this;
+    }
+
+    public int getFrom() {
+        return from;
+    }
+
+    public InnerHitBuilder setFrom(int from) {
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java]: I wonder if we should add validation check to these setter methods like we do in the query builders? So for this one check that the value is >= 0 and for setName() make sure the name passed in is not null.

File: core/src/main/java/org/elasticsearch/index/query/support/InnerHitsBuilder.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    public InnerHitsBuilder() {
+        this.innerHitsBuilders = new HashMap<>();
+    }
+
+    public InnerHitsBuilder(Map<String, InnerHitBuilder> innerHitsBuilders) {
+        this.innerHitsBuilders = innerHitsBuilders;
+    }
+
+    public InnerHitsBuilder addInnerHit(String name, InnerHitBuilder builder) {
+        this.innerHitsBuilders.put(name, builder.setName(name));
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitsBuilder.java]: Maybe check name and builder are not null before adding them to the map?

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/tophits/TopHitsAggregatorFactory.java [hunk 12]
```diff
@@ -79,6 +79,7 @@ public TopHitsAggregatorFactory(String name, Type type, int from, int size, bool
     public Aggregator createInternal(Aggregator parent, boolean collectsFromSingleBucket, List<PipelineAggregator> pipelineAggregators,
             Map<String, Object> metaData) throws IOException {
         SubSearchContext subSearchContext = new SubSearchContext(context.searchContext());
+        subSearchContext.parsedQuery(context.searchContext().parsedQuery());
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/tophits/TopHitsAggregatorFactory.java]: is this related to the inner hits changes? I have no problem with it,I'm just curious if the top_hits somehow ties in with inner_hits

File: core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java [hunk 13]
```diff
@@ -0,0 +1,610 @@
...
+        }
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeOptionalString(name);
+        out.writeOptionalString(nestedPath);
+        out.writeOptionalString(parentChildType);
+        out.writeVLong(from);
+        out.writeVLong(size);
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java]: I just saw from/size get read as VInt, maybe better to write them as such here.

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/tophits/TopHitsAggregatorFactory.java (hunk 12 above)
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/tophits/TopHitsAggregatorFactory.java]: Actually don't worry, I saw below this that you had to change SubSearchContext and thats why this changed

**Comment by colings86** [comment]
@martijnvg I left a few minor comments but otherwise it LGTM

File: core/src/main/java/org/elasticsearch/index/query/support/InnerHitsBuilder.java (hunk 1 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitsBuilder.java]: Could you also check for null here?

File: core/src/main/java/org/elasticsearch/search/SearchService.java [hunk 14]
```diff
@@ -754,20 +756,17 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
...
-                innerHitsParser.nextToken();
-                this.elementParsers.get("inner_hits").parse(innerHitsParser, context);
-            } catch (Exception e) {
-                String sSource = "_na_";
+            for (Map.Entry<String, InnerHitBuilder> entry : source.innerHits().getInnerHitsBuilders().entrySet()) {
                 try {
-                    sSource = source.toString();
-                } catch (Throwable e1) {
-                    // ignore
+                    queryShardContext.reset();
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/SearchService.java]: I'm curious about why this reset (and the one in the finally block) are done. I saw this pattern in other places, just trying to figure out why this is done. In case its important, maybe worth a comment?

File: core/src/main/java/org/elasticsearch/search/fetch/innerhits/InnerHitsContext.java [hunk 15]
```diff
@@ -70,59 +70,49 @@ public InnerHitsContext() {
...
 
-        @Override
-        public ParsedQuery parsedQuery() {
-            return query;
+        public String getName() {
+            return name;
         }
 
-        public abstract TopDocs topDocs(SearchContext context, FetchSubPhase.HitContext hitContext) throws IOException;
+        public void setName(String name) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/fetch/innerhits/InnerHitsContext.java]: Should the name be mandatory, e.g. constructor arg?

File: core/src/main/java/org/elasticsearch/search/fetch/innerhits/InnerHitsFetchSubPhase.java [hunk 16]
```diff
@@ -28,36 +28,28 @@
...
 
     @Inject
-    public InnerHitsFetchSubPhase(FetchSourceParseElement sourceParseElement, HighlighterParseElement highlighterParseElement, FieldDataFieldsParseElement fieldDataFieldsParseElement, ScriptFieldsParseElement scriptFieldsParseElement) {
-        parseElements = singletonMap("inner_hits", new InnerHitsParseElement(sourceParseElement, highlighterParseElement,
-                fieldDataFieldsParseElement, scriptFieldsParseElement));
+    public InnerHitsFetchSubPhase() {
     }
 
     @Override
     public Map<String, ? extends SearchParseElement> parseElements() {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/fetch/innerhits/InnerHitsFetchSubPhase.java]: Can you add a short comment here why this doesn't have a parse element anymore?

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/tophits/TopHitsAggregatorFactory.java (hunk 12 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/tophits/TopHitsAggregatorFactory.java]: Because of this change, inner hits and top hits now share more code.

File: core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java (hunk 13 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java]: good catch! I've changed this.

**Comment by cbuescher** [comment]
@martijnvg I did another quick round, left a few minor suggestions and one question for my own understanding.

File: core/src/main/java/org/elasticsearch/search/SearchService.java (hunk 14 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/search/SearchService.java]: This is the same logic in `QueryShardContext#toQuery()` where we reset also twice. Personally I think a reset at the end is sufficient, but I kept the logic consistent with this method.

The reason we need to invoke reset at all here is because inner hits modifies the `QueryShardContext#nestedScope`, so we need to reset at the end.

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/SearchService.java]: Thansk for the explanation, I've been looking at resets in shard contexts recently and this helps my understanding.

**Comment by martijnvg** [comment]
@cbuescher @colings86 I've updated the PR.

File: core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java [hunk 17]
```diff
@@ -202,7 +200,7 @@ public String getName() {
     }
 
     public InnerHitBuilder setName(String name) {
-        this.name = name;
+        this.name = Objects.requireNonNull(name);
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java]: Is it valid for an inner hits builder to have no name, nestedPath and/or parentChildType set? If not then maybe we should remove the setters for these three fields and have them instead set in the constructor so that it is not possible to construct an InnerHitsBuilder that is not valid?

Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java]: Name is optional, if the name isn't specified it will default to either nestedPath or parentChildType.
I'll add a constructor that accepts either a nestedPath or a parentChildType.

Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java]: scratch that last line... the reason I didn't go with constructor arguments was because: the nested path / parent child type is known while parsing the xcontent for top level inner hits. After we have removed top level inner hits we can make this change easily.

(also name is just a label for reading back the inner hits in the response, lets say you have two nested queries for the same path that have inner hits enabled)

**Comment by cbuescher** [comment]
Thanks, looked at the last changes, LGTM.
